app/controllers/users/backup_code_setup_controller.rb

```diff
@@ -0,0 +1,47 @@
+module Users
+  class BackupCodeSetupController < ApplicationController
+    before_action :authenticate_user!
+    before_action :confirm_two_factor_authenticated, if: :two_factor_enabled?
+
+    def index
+      generate_codes
+      result = BackupCodeSetupForm.new(current_user).submit
+      analytics.track_event(Analytics::BACKUP_CODE_SETUP_VISIT, result.to_h)
+      mark_user_as_fully_authenticated
+    end
+
+    def create
+      analytics.track_event(Analytics::BACKUP_CODE_CREATED)
+      generator.save(user_session[:backup_codes])
```

Comment:
How would the backup_codes not be in session? The GET puts them there.

Reply:
That assumes that the client did a GET first. Never trust the client. :-)

Reply:
What I meant is why are we trying to provide a great experience for hackers or pen testers?  Then again I've cleaned up a multitude of 500's for that very same thing so I see your point.

Reply:
Yup. Not trying to give a great experience to someone trying to hack us. Just trying to prevent the 500s mainly because uncontrolled failures can leave us open to other problems down the road (maybe not in all cases, but making sure failure is managed in a safe way is a good habit).